// @scrypted/auth-fetch/dist/packages/auth-fetch/src/auth-fetch.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAuthFetch = void 0;
const fetch_1 = require("../../../server/src/fetch");
async function getAuth(options, url, method) {
    if (!options.credential)
        return;
    const { BASIC, DIGEST, parseWWWAuthenticateHeader } = await Promise.resolve().then(() => __importStar(require('http-auth-utils')));
    const credential = options.credential;
    const { digest, basic } = credential;
    if (digest) {
        credential.count ||= 0;
        ++credential.count;
        const nc = ('00000000' + credential.count).slice(-8);
        const cnonce = [...Array(24)].map(() => Math.floor(Math.random() * 16).toString(16)).join('');
        const parsedURL = new URL(url);
        const uri = parsedURL.pathname + parsedURL.search;
        const { DIGEST, buildAuthorizationHeader } = await Promise.resolve().then(() => __importStar(require('http-auth-utils')));
        const response = DIGEST.computeHash({
            username: options.credential.username,
            password: options.credential.password,
            method,
            uri,
            nc,
            cnonce,
            algorithm: 'MD5',
            qop: digest.data.qop,
            ...digest.data,
        });
        const header = buildAuthorizationHeader(DIGEST, {
            username: options.credential.username,
            uri,
            nc,
            cnonce,
            algorithm: digest.data.algorithm,
            qop: digest.data.qop,
            response,
            ...digest.data,
        });
        return header;
    }
    else if (basic) {
        const { BASIC, buildAuthorizationHeader } = await Promise.resolve().then(() => __importStar(require('http-auth-utils')));
        const header = buildAuthorizationHeader(BASIC, {
            username: options.credential.username,
            password: options.credential.password,
        });
        return header;
    }
}
function createAuthFetch(h, parser) {
    const authHttpFetch = async (options) => {
        const method = (0, fetch_1.getFetchMethod)(options);
        const headers = (0, fetch_1.createHeadersArray)(options.headers);
        options.headers = headers;
        (0, fetch_1.setDefaultHttpFetchAccept)(headers, options.responseType);
        const initialHeader = await getAuth(options, options.url, method);
        // try to provide an authorization if a session exists, but don't override Authorization if provided already.
        // 401 will trigger a proper auth.
        if (initialHeader && !(0, fetch_1.hasHeader)(headers, 'Authorization'))
            (0, fetch_1.setHeader)(headers, 'Authorization', initialHeader);
        const controller = new AbortController();
        options.signal?.addEventListener('abort', () => controller.abort(options.signal?.reason));
        const initialResponse = await h({
            ...options,
            signal: controller.signal,
            // need to intercept the status code to check for 401.
            // all other status codes will be handled according to the initial request options.
            checkStatusCode(statusCode) {
                // can handle a 401 if an credential is provided.
                // however, not providing a credential is also valid, and should
                // fall through to the normal response handling which may be interested
                // in the 401 response.
                if (statusCode === 401 && options.credential)
                    return true;
                if (options?.checkStatusCode === undefined || options?.checkStatusCode) {
                    const checker = typeof options?.checkStatusCode === 'function' ? options.checkStatusCode : fetch_1.checkStatus;
                    return checker(statusCode);
                }
                return true;
            },
            responseType: 'readable',
        });
        // if it's not a 401, just return the response.
        if (initialResponse.statusCode !== 401) {
            return {
                ...initialResponse,
                body: await parser(initialResponse.body, options.responseType),
            };
        }
        let authenticateHeaders = initialResponse.headers.get('www-authenticate');
        if (!authenticateHeaders)
            throw new Error('Did not find WWW-Authenticate header.');
        if (typeof authenticateHeaders === 'string')
            authenticateHeaders = [authenticateHeaders];
        const { BASIC, DIGEST, parseWWWAuthenticateHeader } = await Promise.resolve().then(() => __importStar(require('http-auth-utils')));
        const parsedHeaders = authenticateHeaders.map(h => parseWWWAuthenticateHeader(h));
        const digest = parsedHeaders.find(p => p.type === 'Digest');
        const basic = parsedHeaders.find(p => p.type === 'Basic');
        const credential = options.credential;
        credential.digest = digest;
        credential.basic = basic;
        if (!digest && !basic)
            throw new Error(`Unknown WWW-Authenticate type: ${parsedHeaders[0]?.type}`);
        const header = await getAuth(options, options.url, method);
        if (header)
            (0, fetch_1.setHeader)(headers, 'Authorization', header);
        return h(options);
    };
    return authHttpFetch;
}
exports.createAuthFetch = createAuthFetch;
