// @scrypted/auth-fetch/dist/server/src/fetch/http-fetch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.httpFetch = exports.httpFetchParseIncomingMessage = exports.getHttpFetchParser = void 0;
const _1 = require(".");
async function readMessageBuffer(response) {
    const buffers = [];
    response.on('data', buffer => buffers.push(buffer));
    const { once } = require('events');
    await once(response, 'end');
    return Buffer.concat(buffers);
}
const TextParser = {
    async parse(message) {
        return (await readMessageBuffer(message)).toString();
    }
};
const JSONParser = {
    async parse(message) {
        return JSON.parse((await readMessageBuffer(message)).toString());
    }
};
const BufferParser = {
    async parse(message) {
        return readMessageBuffer(message);
    }
};
const StreamParser = {
    async parse(message) {
        return message;
    }
};
function getHttpFetchParser(responseType) {
    switch (responseType) {
        case 'json':
            return JSONParser;
        case 'text':
            return TextParser;
        case 'readable':
            return StreamParser;
    }
    return BufferParser;
}
exports.getHttpFetchParser = getHttpFetchParser;
function httpFetchParseIncomingMessage(readable, responseType) {
    return getHttpFetchParser(responseType).parse(readable);
}
exports.httpFetchParseIncomingMessage = httpFetchParseIncomingMessage;
async function httpFetch(options) {
    const headers = (0, _1.createHeadersArray)(options.headers);
    (0, _1.setDefaultHttpFetchAccept)(headers, options.responseType);
    const { once } = require('events');
    const { PassThrough, Readable } = require('stream');
    const { http, https } = require('follow-redirects');
    const { url } = options;
    const isSecure = url.toString().startsWith('https:');
    const proto = isSecure ? https : http;
    let { body } = options;
    if (body && !(body instanceof Readable)) {
        const newBody = new PassThrough();
        newBody.write(Buffer.from((0, _1.createStringOrBufferBody)(headers, body)));
        newBody.end();
        body = newBody;
    }
    let controller;
    let timeout;
    if (options.timeout) {
        controller = new AbortController();
        timeout = setTimeout(() => controller.abort(), options.timeout);
        options.signal?.addEventListener('abort', () => controller.abort(options.signal?.reason));
    }
    const signal = controller?.signal || options.signal;
    signal?.addEventListener('abort', () => request.destroy(new Error(options.signal?.reason || 'abort')));
    const nodeHeaders = {};
    for (const [k, v] of headers) {
        if (nodeHeaders[k]) {
            nodeHeaders[k].push(v);
        }
        else {
            nodeHeaders[k] = [v];
        }
    }
    const request = proto.request(url, {
        method: (0, _1.getFetchMethod)(options),
        rejectUnauthorized: options.rejectUnauthorized,
        family: options.family,
        headers: nodeHeaders,
        signal,
        timeout: options.timeout,
    });
    if (body)
        body.pipe(request);
    else
        request.end();
    try {
        const [response] = await once(request, 'response');
        if (options?.checkStatusCode === undefined || options?.checkStatusCode) {
            try {
                const checker = typeof options?.checkStatusCode === 'function' ? options.checkStatusCode : _1.checkStatus;
                if (!response.statusCode || !checker(response.statusCode))
                    throw new Error(`http response statusCode ${response.statusCode}`);
            }
            catch (e) {
                readMessageBuffer(response).catch(() => { });
                throw e;
            }
        }
        const incomingHeaders = new Headers();
        for (const [k, v] of Object.entries(response.headers)) {
            for (const vv of (typeof v === 'string' ? [v] : v)) {
                incomingHeaders.append(k, vv);
            }
        }
        return {
            statusCode: response.statusCode,
            headers: incomingHeaders,
            body: await httpFetchParseIncomingMessage(response, options.responseType),
        };
    }
    finally {
        clearTimeout(timeout);
    }
}
exports.httpFetch = httpFetch;
