// @scrypted/auth-fetch/dist/server/src/fetch/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.domFetch = exports.domFetchParseIncomingMessage = exports.createStringOrBufferBody = exports.createHeadersArray = exports.setDefaultHttpFetchAccept = exports.setHeader = exports.removeHeader = exports.hasHeader = exports.getHttpFetchAccept = exports.getFetchMethod = exports.checkStatus = exports.fetchStatusCodeOk = void 0;
function fetchStatusCodeOk(statusCode) {
    return statusCode >= 200 && statusCode <= 299;
}
exports.fetchStatusCodeOk = fetchStatusCodeOk;
function checkStatus(statusCode) {
    if (!fetchStatusCodeOk(statusCode))
        throw new Error(`http response statusCode ${statusCode}`);
    return true;
}
exports.checkStatus = checkStatus;
function getFetchMethod(options) {
    const method = options.method || (options.body ? 'POST' : 'GET');
    return method;
}
exports.getFetchMethod = getFetchMethod;
function getHttpFetchAccept(responseType) {
    switch (responseType) {
        case 'json':
            return 'application/json';
        case 'text':
            return 'text/plain';
    }
    return;
}
exports.getHttpFetchAccept = getHttpFetchAccept;
function hasHeader(headers, key) {
    key = key.toLowerCase();
    return headers.find(([k]) => k.toLowerCase() === key);
}
exports.hasHeader = hasHeader;
function removeHeader(headers, key) {
    key = key.toLowerCase();
    const filteredHeaders = headers.filter(([headerKey, _]) => headerKey.toLowerCase() !== key);
    headers.length = 0;
    filteredHeaders.forEach(header => headers.push(header));
}
exports.removeHeader = removeHeader;
function setHeader(headers, key, value) {
    removeHeader(headers, key);
    headers.push([key, value]);
}
exports.setHeader = setHeader;
function setDefaultHttpFetchAccept(headers, responseType) {
    if (hasHeader(headers, 'Accept'))
        return;
    const accept = getHttpFetchAccept(responseType);
    if (accept)
        setHeader(headers, 'Accept', accept);
}
exports.setDefaultHttpFetchAccept = setDefaultHttpFetchAccept;
function createHeadersArray(headers) {
    const headersArray = [];
    if (!headers)
        return headersArray;
    if (headers instanceof Headers) {
        for (const [k, v] of headers.entries()) {
            headersArray.push([k, v]);
        }
        return headersArray;
    }
    if (headers instanceof Array) {
        for (const [k, v] of headers) {
            headersArray.push([k, v]);
        }
        return headersArray;
    }
    for (const k of Object.keys(headers)) {
        const v = headers[k];
        headersArray.push([k, v]);
    }
    return headersArray;
}
exports.createHeadersArray = createHeadersArray;
/**
 *
 * @param headers
 * @param body
 * @returns Returns the body and Content-Type header that was set.
 */
function createStringOrBufferBody(headers, body) {
    let contentType;
    if (typeof body === 'object' && !ArrayBuffer.isView(body)) {
        body = JSON.stringify(body);
        contentType = 'application/json';
    }
    else if (typeof body === 'string') {
        contentType = 'text/plain';
    }
    if (contentType && !hasHeader(headers, 'Content-Type'))
        setHeader(headers, 'Content-Type', contentType);
    if (!hasHeader(headers, 'Content-Length')) {
        body = Buffer.from(body);
        setHeader(headers, 'Content-Length', body.length.toString());
    }
    return body;
}
exports.createStringOrBufferBody = createStringOrBufferBody;
async function domFetchParseIncomingMessage(response, responseType) {
    switch (responseType) {
        case 'json':
            return response.json();
        case 'text':
            return response.text();
        case 'readable':
            return response;
    }
    return new Uint8Array(await response.arrayBuffer());
}
exports.domFetchParseIncomingMessage = domFetchParseIncomingMessage;
async function domFetch(options) {
    const headers = createHeadersArray(options.headers);
    setDefaultHttpFetchAccept(headers, options.responseType);
    let { body } = options;
    if (body && !(body instanceof ReadableStream)) {
        body = createStringOrBufferBody(headers, body);
    }
    let controller;
    let timeout;
    if (options.timeout) {
        controller = new AbortController();
        timeout = setTimeout(() => controller.abort(), options.timeout);
        options.signal?.addEventListener('abort', () => controller.abort(options.signal?.reason));
    }
    try {
        const { url } = options;
        const response = await fetch(url, {
            method: getFetchMethod(options),
            credentials: options.withCredentials ? 'include' : undefined,
            headers,
            signal: controller?.signal || options.signal,
            body,
        });
        if (options?.checkStatusCode === undefined || options?.checkStatusCode) {
            try {
                const checker = typeof options?.checkStatusCode === 'function' ? options.checkStatusCode : checkStatus;
                if (!checker(response.status))
                    throw new Error(`http response statusCode ${response.status}`);
            }
            catch (e) {
                response.arrayBuffer().catch(() => { });
                throw e;
            }
        }
        return {
            statusCode: response.status,
            headers: response.headers,
            body: await domFetchParseIncomingMessage(response, options.responseType),
        };
    }
    finally {
        clearTimeout(timeout);
    }
}
exports.domFetch = domFetch;
