// @scrypted/sdk/dist/src/bin/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deploy = deploy;
exports.debug = debug;
exports.getDefaultWebpackConfig = getDefaultWebpackConfig;
const auth_fetch_1 = require("@scrypted/auth-fetch");
const process_1 = __importDefault(require("process"));
const path_1 = __importDefault(require("path"));
const fs_1 = __importDefault(require("fs"));
function getUserHome() {
    const ret = process_1.default.env[(process_1.default.platform == 'win32') ? 'USERPROFILE' : 'HOME'];
    if (!ret)
        throw new Error('Neither USERPROFILE or HOME are defined.');
    return ret;
}
const scryptedHome = path_1.default.join(getUserHome(), '.scrypted');
const loginPath = path_1.default.join(scryptedHome, 'login.json');
function getLogin(ip) {
    let login;
    try {
        login = JSON.parse(fs_1.default.readFileSync(loginPath).toString());
    }
    catch {
        login = {};
    }
    const ipLogin = login[ip];
    return {
        username: ipLogin?.username || '',
        password: ipLogin?.token || '',
    };
}
function showLoginError() {
    console.error('Authorization required. Please log in with the following:');
    console.error('     npx scrypted login [ip]');
}
function toIpAndPort(ip) {
    if (ip.indexOf(':') === -1)
        ip += ':10443';
    console.log(ip);
    return ip;
}
async function deploy(debugHost, noRebind) {
    debugHost = toIpAndPort(debugHost);
    let out;
    if (process_1.default.env.NODE_ENV === 'production')
        out = path_1.default.resolve(process_1.default.cwd(), 'dist');
    else
        out = path_1.default.resolve(process_1.default.cwd(), 'out');
    const outFilename = 'plugin.zip';
    const main = path_1.default.resolve(out, outFilename);
    if (!fs_1.default.existsSync(main)) {
        console.error('npm run scrypted-webpack to build a webpack bundle for Scrypted.');
        throw new Error(`Missing webpack bundle: ${main}`);
    }
    const packageJsonPath = path_1.default.resolve(process_1.default.cwd(), 'package.json');
    const packageJson = JSON.parse(fs_1.default.readFileSync(packageJsonPath).toString());
    const npmPackage = packageJson.name || '';
    const rebindQuery = noRebind ? 'no-rebind' : '';
    const deployUrl = `https://${debugHost}/web/component/script/deploy?${rebindQuery}&npmPackage=${npmPackage}`;
    const setupUrl = `https://${debugHost}/web/component/script/setup?${rebindQuery}&npmPackage=${npmPackage}`;
    const fileContents = fs_1.default.readFileSync(main);
    console.log(`deploying to ${debugHost}`);
    const auth = getLogin(debugHost);
    try {
        await (0, auth_fetch_1.httpFetch)({
            url: setupUrl,
            method: 'POST',
            body: JSON.stringify(packageJson),
            headers: {
                ...basicAuthHeaders(auth.username, auth.password),
                'Content-Type': 'application/json',
            },
            timeout: 10000,
            rejectUnauthorized: false,
            checkStatusCode(statusCode) {
                if (statusCode === 401) {
                    showLoginError();
                }
                return statusCode >= 200 && statusCode < 300;
            },
        });
        console.log(`configured ${debugHost}`);
        await (0, auth_fetch_1.httpFetch)({
            url: deployUrl,
            method: 'POST',
            body: fileContents,
            timeout: 10000,
            rejectUnauthorized: false,
            headers: {
                ...basicAuthHeaders(auth.username, auth.password),
                'Content-Type': 'application/zip',
            },
        });
        console.log(`deployed to ${debugHost}`);
    }
    catch (err) {
        const error = err;
        console.error(error.message);
        throw error;
    }
}
function basicAuthHeaders(username, password) {
    return {
        'Authorization': `Basic ${Buffer.from(username + ":" + password).toString('base64')}`
    };
}
async function debug(debugHost, entryPoint) {
    debugHost = toIpAndPort(debugHost);
    const packageJsonPath = path_1.default.resolve(process_1.default.cwd(), 'package.json');
    const packageJson = JSON.parse(fs_1.default.readFileSync(packageJsonPath).toString());
    const npmPackage = packageJson.name || '';
    const debugUrl = `https://${debugHost}/web/component/script/debug?npmPackage=${npmPackage}`;
    console.log(`initiating debugger on ${debugHost}`);
    const auth = getLogin(debugHost);
    try {
        await (0, auth_fetch_1.httpFetch)({
            url: debugUrl,
            method: 'POST',
            headers: basicAuthHeaders(auth.username, auth.password),
            timeout: 10000,
            rejectUnauthorized: false,
        });
        console.log(`debugger ready on ${debugHost}`);
    }
    catch (err) {
        const error = err;
        console.error(error.message);
        throw error;
    }
}
function getDefaultWebpackConfig(name) {
    return require(path_1.default.resolve(__dirname, `../../../${name}`));
}
