// @scrypted/sdk/dist/src/bin/scrypted-webpack.js
#! /usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
try {
    require('adm-zip');
}
catch {
    throw new Error('Please "npm install" in the "sdk" directory.');
}
const path_1 = __importDefault(require("path"));
const process_1 = __importDefault(require("process"));
const fs_1 = __importDefault(require("fs"));
const os_1 = __importDefault(require("os"));
const adm_zip_1 = __importDefault(require("adm-zip"));
const rimraf_1 = require("rimraf");
const webpack_1 = __importDefault(require("webpack"));
const tmp_1 = __importDefault(require("tmp"));
const child_process_1 = __importDefault(require("child_process"));
const events_1 = require("events");
const cwd = process_1.default.cwd();
let out;
if (process_1.default.env.NODE_ENV === 'production')
    out = path_1.default.resolve(cwd, 'dist');
else
    out = path_1.default.resolve(cwd, 'out');
if (fs_1.default.existsSync(path_1.default.resolve(cwd, 'src/main.py'))) {
    const resolved = path_1.default.resolve(cwd, 'src');
    const zip = new adm_zip_1.default();
    const readme = path_1.default.join(cwd, 'README.md');
    if (fs_1.default.existsSync(readme)) {
        zip.addLocalFile(readme);
    }
    zip.addLocalFolder(resolved);
    const sdk = path_1.default.join(__dirname, '../../../types/scrypted_python/scrypted_sdk');
    zip.addLocalFolder(sdk, 'scrypted_sdk', filename => !filename.endsWith('.pyc'));
    const zipfs = path_1.default.join(cwd, 'fs');
    if (fs_1.default.existsSync(zipfs))
        zip.addLocalFolder(zipfs, 'fs');
    zip.writeZip(path_1.default.join(out, 'plugin.zip'));
    process_1.default.exit(0);
}
const packageJson = JSON.parse(fs_1.default.readFileSync(path_1.default.join(cwd, 'package.json'), 'utf8'));
const interfaceDescriptors = packageJson.scrypted?.interfaceDescriptors;
const optionalDependencies = Object.keys(packageJson.optionalDependencies || {});
if (packageJson.scrypted?.babel) {
    process_1.default.env.SCRYPTED_WEBPACK_BABEL = 'true';
}
const defaultMainNodeJs = 'main.nodejs.js';
const entries = [];
if (packageJson.exports) {
    for (const [key, value] of Object.entries(packageJson.exports)) {
        entries.push({
            filename: key,
            output: value,
        });
    }
}
else {
    for (const search of ['src/main.js', 'src/main.ts']) {
        const resolved = path_1.default.resolve(cwd, search);
        if (fs_1.default.existsSync(resolved)) {
            entries.push({
                filename: search,
                output: defaultMainNodeJs,
            });
            break;
        }
    }
}
const nodeWebpackConfig = 'webpack.nodejs.config.js';
if (!entries?.length) {
    console.warn('unable to locate src/main.js or src/main.ts');
    console.warn('if a custom webpack config is used, will fall back to an entry configured there');
    entries.push(undefined);
}
const zip = new adm_zip_1.default();
const readme = path_1.default.join(cwd, 'README.md');
if (fs_1.default.existsSync(readme)) {
    let readmeText = fs_1.default.readFileSync(readme).toString();
    const changelog = path_1.default.join(cwd, 'CHANGELOG.md');
    if (fs_1.default.existsSync(changelog)) {
        readmeText += '\n\n\n<br/><br/>' + fs_1.default.readFileSync(changelog).toString();
    }
    zip.addFile('README.md', Buffer.from(readmeText));
}
const NODE_PATH = path_1.default.resolve(__dirname, '..', '..', '..', 'node_modules');
process_1.default.env.NODE_PATH = NODE_PATH;
require('module').Module._initPaths();
async function rollup() {
    if (out)
        (0, rimraf_1.rimrafSync)(out);
    let rollupCmd = path_1.default.resolve(cwd, 'node_modules/.bin/rollup');
    if (!fs_1.default.existsSync(rollupCmd)) {
        rollupCmd = path_1.default.resolve(cwd, 'node_modules/@scrypted/sdk/node_modules/.bin/rollup');
    }
    if (os_1.default.platform().startsWith('win')) {
        rollupCmd += '.cmd';
    }
    const cp = child_process_1.default.spawn(rollupCmd, [
        '--config', path_1.default.resolve(__dirname, '../../../rollup.nodejs.config.mjs'),
    ], {
        stdio: 'inherit',
    });
    await (0, events_1.once)(cp, 'exit');
    if (cp.exitCode)
        throw new Error('rollup failed');
    finishZip();
}
function finishZip() {
    const jsFiles = fs_1.default.readdirSync(out, {
        withFileTypes: true
    }).filter(ft => ft.isFile() && ft.name.endsWith('.js')).map(ft => ft.name);
    for (const js of jsFiles) {
        zip.addLocalFile(path_1.default.join(out, js));
        const sourcemap = path_1.default.join(out, js + '.map');
        if (fs_1.default.existsSync(sourcemap))
            zip.addLocalFile(sourcemap);
        console.log(js);
    }
    const sdkPackageJson = require(path_1.default.join(__dirname, '../../../package.json'));
    zip.addFile('sdk.json', Buffer.from(JSON.stringify({
        version: sdkPackageJson.version,
        interfaceDescriptors,
    })));
    if (packageJson.type === 'module') {
        zip.addFile('package.json', Buffer.from(JSON.stringify({
            type: 'module'
        })));
    }
    const zipfs = path_1.default.join(cwd, 'fs');
    if (fs_1.default.existsSync(zipfs))
        zip.addLocalFolder(zipfs, 'fs');
    zip.writeZip(path_1.default.join(out, 'plugin.zip'));
}
async function pack() {
    if (out)
        (0, rimraf_1.rimrafSync)(out);
    await new Promise((resolve, reject) => {
        let webpackConfig;
        const customWebpackConfig = path_1.default.resolve(cwd, nodeWebpackConfig);
        const defaultWebpackConfig = path_1.default.resolve(__dirname, '..', '..', '..', nodeWebpackConfig);
        if (fs_1.default.existsSync(customWebpackConfig)) {
            webpackConfig = customWebpackConfig;
        }
        else {
            webpackConfig = defaultWebpackConfig;
        }
        process_1.default.env.SCRYPTED_DEFAULT_WEBPACK_CONFIG = defaultWebpackConfig;
        const webpackEntries = {};
        const config = require(webpackConfig);
        for (const entry of entries) {
            const normalizedEntry = entry || {
                filename: (typeof config?.entry === 'object' ? config.entry?.main : config.entry) || '',
                output: defaultMainNodeJs,
            };
            if (!normalizedEntry?.filename) {
                console.error("no main.ts or main.js was found, and webpack config does not supply an entry file.");
                console.error(normalizedEntry?.filename);
                throw new Error();
            }
            const main = path_1.default.resolve(cwd, normalizedEntry.filename);
            if (!fs_1.default.existsSync(main)) {
                console.error("entry file specified in webpack config does not exist");
                throw new Error();
            }
            webpackEntries[normalizedEntry.output] = main;
        }
        config.entry = webpackEntries;
        config.output = config.output || {};
        config.output.filename = '[name]';
        config.output.path = out;
        config.resolve = config.resolve || {};
        config.resolve.alias = config.resolve.alias || {};
        for (const opt of optionalDependencies) {
            const t = tmp_1.default.tmpNameSync({
                postfix: '.js',
            });
            fs_1.default.writeFileSync(t, `
                        const e = __non_webpack_require__('${opt}');
                        module.exports = e;
                    `);
            config.resolve.alias[opt] = t;
        }
        (0, webpack_1.default)(config, (err, stats) => {
            if (err)
                return reject(err);
            if (stats?.hasErrors()) {
                console.error(stats.toJson()?.errors);
                return reject(new Error('webpack failed'));
            }
            resolve();
        });
    });
    finishZip();
}
(packageJson.scrypted?.rollup ? rollup : pack)()
    .catch(e => process_1.default.nextTick(() => {
    console.error(e);
    throw new Error(e);
}));
